server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -219,14 +222,37 @@ public CachingClusteredClient(
     final List<TimelineObjectHolder<String, ServerSelector>> filteredServersLookup =
         toolChest.filterSegments(query, serversLookup);
 
+    Map<String, RangeSet<String>> dimensionRangeMap = Maps.newHashMap();
+    DimFilter filter = query.getFilter();
+
+    // Filter unneeded chunks based on partition dimension
     for (TimelineObjectHolder<String, ServerSelector> holder : filteredServersLookup) {
       for (PartitionChunk<ServerSelector> chunk : holder.getObject()) {
         ServerSelector selector = chunk.getObject();
-        final SegmentDescriptor descriptor = new SegmentDescriptor(
-            holder.getInterval(), holder.getVersion(), chunk.getChunkNumber()
-        );
+        boolean include = true;
+
+        if (filter != null) {
+          Map<String, RangeSet<String>> domain = selector.getSegment().getShardSpec().getDomain();
+          for (Map.Entry<String, RangeSet<String>> entry : domain.entrySet()) {
+            if (dimensionRangeMap.get(entry.getKey()) == null) {
```

Comment:
Could also stash the result of `dimensionRangeMap.get(entry.getKey())` in a variable to avoid extra lookups